File: Java/DailyLeetCode/minObstacles.java
import java.util.*;
class Solution {
    public int minimumObstacles(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        
        int[][] dir = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        
        Deque<int[]> queue = new ArrayDeque<>();
        queue.add(new int[]{0, 0, 0}); 
        int[][] minObstacles = new int[rows][cols];
        for (int[] row : minObstacles) {
            Arrays.fill(row, Integer.MAX_VALUE);
        }
        minObstacles[0][0] = 0;
        
        while (!queue.isEmpty()) {
            int[] currentPosition = queue.poll();
            int currentRow = currentPosition[0], currentCol = currentPosition[1], obstaclesRemoved = currentPosition[2];
            
            if (currentRow == rows - 1 && currentCol == cols - 1) {
                return obstaclesRemoved; 
            }
            
            for (int[] direction : dir) {
                int newRow = currentRow + direction[0];
                int newCol = currentCol + direction[1];
                
                if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols) {
                    int newObstaclesRemoved = obstaclesRemoved + grid[newRow][newCol];
                    
                    if (newObstaclesRemoved < minObstacles[newRow][newCol]) {
                        minObstacles[newRow][newCol] = newObstaclesRemoved;
                        
                        if (grid[newRow][newCol] == 0) {
                            queue.addFirst(new int[]{newRow, newCol, newObstaclesRemoved});
                        } else {
                            queue.addLast(new int[]{newRow, newCol, newObstaclesRemoved});
                        }
                    }
                }
            }
        }
        
        return -1;
    }
}

File: Python/DailyLeetCode/minObstacles.py
from collections import deque
from typing import List

class Solution:
    def minimumObstacles(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])

        reachable = [[True] * cols for _ in range(rows)]
        reachable[0][0] = False

        directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

        queue = deque([(0, 0, 0)])
        while queue:
            current_row, current_col, current_distance = queue.popleft()

            for delta_row, delta_col in directions:
                next_row = current_row + delta_row
                next_col = current_col + delta_col

                if 0 <= next_row < rows and 0 <= next_col < cols and reachable[next_row][next_col]:
                    reachable[next_row][next_col] = False

                    if next_row == rows - 1 and next_col == cols - 1:
                        return current_distance
                    
                    if grid[next_row][next_col] == 0:
                        queue.appendleft((next_row, next_col, current_distance))
                    else:
                        queue.append((next_row, next_col, current_distance + 1))
